ClassFieldTheory/IsNonarchimedeanLocalField/Unramified.lean

```diff
@@ -0,0 +1,320 @@
+deriving Field, Algebra K, FiniteDimensional K
+
+variable (K : Type*) [Field K] [ValuativeRel K] [TopologicalSpace K] [IsNonarchimedeanLocalField K]
+variable (n : ℕ)
+
+instance : ValuativeRel (UnramifiedExtension K n) :=
+  (isNonarchimedeanLocalField_of_finiteDimensional K _).choose
+
+instance : ValuativeExtension K (UnramifiedExtension K n) :=
+  (isNonarchimedeanLocalField_of_finiteDimensional K _).choose_spec.choose
+
+instance : TopologicalSpace (UnramifiedExtension K n) :=
+  (isNonarchimedeanLocalField_of_finiteDimensional K _).choose_spec.choose_spec.choose
+
+instance : IsNonarchimedeanLocalField (UnramifiedExtension K n) :=
+  (isNonarchimedeanLocalField_of_finiteDimensional K _).choose_spec.choose_spec.choose_spec
+
+instance UnramifiedExtension.isSplittingField :
+    IsSplittingField K (UnramifiedExtension K n) (X ^ (Nat.card 𝓀[K] ^ n - 1) - 1) :=
+  .splittingField _
+
+theorem UnramifiedExtension.splits :
+    (X ^ (Nat.card 𝓀[K] ^ n - 1) - 1 : (UnramifiedExtension K n)[X]).Splits (.id _) := by
+  have := (UnramifiedExtension.isSplittingField K n).splits
+  simpa using (splits_id_iff_splits _).mpr this
+
+theorem UnramifiedExtension.factors :
+    (X ^ (Nat.card 𝓀[K] ^ n - 1) - 1 : (UnramifiedExtension K n)[X]).Factors := by
+  simpa [Splits] using splits K n
+
+open UnramifiedExtension
+
+section zero
+
+theorem finrank_unramifiedExtension_zero : Module.finrank K (UnramifiedExtension K 0) = 1 := by
+  have := UnramifiedExtension.isSplittingField K 0
+  rw [pow_zero, Nat.sub_self, pow_zero, sub_self, isSplittingField_iff_intermediateField,
+    rootSet_zero, IntermediateField.adjoin_empty,
+    IntermediateField.bot_eq_top_iff_finrank_eq_one] at this
+  exact this.2
+
+instance : Subsingleton (Subalgebra K (UnramifiedExtension K 0)) :=
+  subsingleton_of_bot_eq_top <| Subalgebra.bot_eq_top_of_finrank_eq_one <|
+    finrank_unramifiedExtension_zero K
+
+instance : Subsingleton (IntermediateField K (UnramifiedExtension K 0)) :=
+  IntermediateField.toSubalgebra_injective.subsingleton
+
+end zero
+
+-- An auxiliary lemma that we might need more than once
+theorem card_residue_pow_sub_one_in_residue_ne_zero {n : ℕ} (hn : n ≠ 0) :
+    ((Nat.card 𝓀[K] ^ n - 1 :) : 𝓀[K]) ≠ 0 := by
+  have hp := Fact.mk <| prime_ringChar K
+  have := ZMod.algebra
+  rw [ne_eq, CharP.cast_eq_zero_iff, Module.natCard_eq_pow_finrank (K := ZMod (ringChar 𝓀[K])),
+    Nat.card_zmod, ← pow_mul, Nat.dvd_sub_iff_right, Nat.dvd_one]
+  · exact hp.out.ne_one
+  · exact one_le_pow_of_one_le' hp.out.one_le _
+  · exact dvd_pow_self _ <| mul_ne_zero Module.finrank_pos.ne' hn
+
+-- An auxiliary lemma that we might need more than once
+theorem card_residue_pow_sub_one_in_integers_ne_zero {n : ℕ} (hn : n ≠ 0) :
+    ((Nat.card 𝓀[K] ^ n - 1 :) : 𝒪[K]) ≠ 0 := by
+  refine ne_zero_of_map (f := algebraMap 𝒪[K] 𝓀[K]) ?_
+  rw [map_natCast]
+  exact card_residue_pow_sub_one_in_residue_ne_zero K hn
+
+-- An auxiliary lemma that we might need more than once
```

Comment:
You need it three times. Why not just say "...that we need more than once"?

Reply:
done